scvi/models/_base.py

```diff
@@ -0,0 +1,792 @@
+        batch_size=128,
+        return_mean: bool = True,
+        return_numpy: Optional[bool] = None,
+    ) -> Union[np.ndarray, pd.DataFrame]:
+        # whatever is here will pass it into cat and batch
+        r"""
+        Returns the normalized (decoded) gene expression.
+
+        This is denoted as :math:`\rho_n` in the scVI paper.
+
+        Parameters
+        ----------
+        adata
+            AnnData object that has been registered with scvi. If `None`, defaults to the
+            AnnData object used to initialize the model.
+        indices
+            Indices of cells in adata to use. If `None`, all cells are used.
+        transform_batch
+            Batch to condition on.
+            If transform_batch is:
+
+            - None, then real observed batch is used
+            - int, then batch transform_batch is used
+        gene_list
+            Return frequencies of expression for a subset of genes.
+            This can save memory when working with large datasets and few genes are
+            of interest.
+        library_size
+            Scale the expression frequencies to a common library size.
+            This allows gene expression levels to be interpreted on a common scale of relevant
+            magnitude.
+        n_samples
+            Get sample scale from multiple samples.
+        batch_size
+            Minibatch size for data loading into model
+        return_mean
+            Whether to return the mean of the samples.
+        return_numpy
+            Return a `np.ndarray` instead of a `pd.DataFrame`. Includes gene
+            names as columns. If either n_samples=1 or return_mean=True, defaults to False.
+            Otherwise, it defaults to True.
+
+        Returns
+        -------
+        - **normalized_expression** - array of normalized expression
+
+        If ``n_samples`` > 1 and ``return_mean`` is False, then the shape is ``(samples, cells, genes)``.
+        Otherwise, shape is ``(cells, genes)``. Return type is ``pd.DataFrame`` unless ``return_numpy`` is True.
+        """
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
+
+        if gene_list is None:
+            gene_mask = slice(None)
+        else:
+            all_genes = _get_var_names_from_setup_anndata(adata)
+            gene_mask = [True if gene in gene_list else False for gene in all_genes]
+
+        if n_samples > 1 and return_mean is False:
+            if return_numpy is False:
+                logger.warning(
+                    "return_numpy must be True if n_samples > 1 and return_mean is False, returning np.ndarray"
+                )
+            return_numpy = True
+        if indices is None:
+            indices = np.arange(adata.n_obs)
+        if library_size == "latent":
+            model_fn = self.model.get_sample_rate
+            scaling = 1
+        else:
+            model_fn = self.model.get_sample_scale
```

Comment:
sounds wierd for a model to use library size as latent and be able to perform DE as it it were not. But it seems handy for new models somehow. Still, it's confusing so this deserves a couple of comments about what these functions are doing and what's happening here